### ISSUE TITLE
OffscreenCanvas and webworker for efficient visualization

### ISSUE THREAD

**Issue by jetic83** [openseadragon/openseadragon#2352]
OffscreenCanvas and WebWorkers are great tools to decouple canvas related actions from the main thread, increasing efficiency and responsiveness.

Would this be an option for OSD as well? How much effort would this be to implement (as an estimate)?

The reason would be e.g. load tiles in the background, render in the background and so on, while the main thread might be busy.

**Comment by iangilman** [comment]
Great question! I wasn't aware of the OffscreenCanvas; thank you for bringing it up! 

We use canvasses in a number of ways in OSD... In addition to the main canvas drawing surface, we store tiles in canvasses and we have a "sketch canvas" that we do pre-drawing work in depending on the scenario. Converting everything but the canvas drawing surface to offscreen canvasses would probably be pretty straightforward. It's hard to say how much of a performance improvement that would be, but probably still worthwhile. 

Converting the main drawing canvas would presumably involve creating a new offscreen canvas and doing all of the tile drawing on it, then copying the result to the onscreen canvas. I have no idea if that would be faster (because the bulk of the drawing is happening offscreen) or slower (because there's an additional draw call to go from the offscreen canvas to the onscreen one); I guess one would just need to try it and see. It certainly would increase our memory requirement, which wouldn't be welcome in some scenarios. 

I'm not sure what we would use WebWorkers for... We want the drawing to happen on the main thread, because our goal is to do it 60 frames a second and shuttling data back and forth between workers will introduce latency. The processing that would be nice to move off the main thread is loading the tiles, but I think the big weight there is decompressing the image files as they load, and I don't think that can be done outside the main thread (unless we want to include our own image decoders, which I'm pretty sure we don't). 

One place WebWorkers do get used is in plugins that manipulate the tiles in various ways (like adjusting their contrast). That's a perfect use for WebWorkers. 

Those are my initial thoughts... What do you think? What did you have in mind? 

Thank again for bringing up this topic!

**Comment by Aiosa** [comment]
IMHO, there are two drawing points: tile canvasses (could be offscreen canvasses, good idea) and the actual animation rendering. Performing animation rendering on an offscreen canvas is not a good idea, because we draw _the animation_ - that's what the main thread is supposed to do. But moving the tile preprocessing to offscreen implementation might be a good thing - I was not aware of this API as well. For efficient visualizations, I am using WebGL (https://rationai-vis.ics.muni.cz/visualization-demo/showcase.php) and we are planning to bring a WebGL renderer to the OSD in a few months, so I guess it replaces the need for offscreen canvas...

**Comment by jetic83** [comment]
Thanks a lot!

I guess webGL would be the better option indeed. 

The only reason I brought this up a actually to make OSD as smooth as possible, without any latency when navigating through the image. This is usually a problem in digital pathology viewers based on OSD (it's never really videolike). One reason is network, but another big reason is rendering. 

I think with the reference to webGL, this ticket can be closed. 

Thanks for discussing!

**Comment by nmzali** [comment]
@iangilman I also wanted to give my two cents on OffscreenCanvas and WebWorkers. I have a personal project that I'm working on exploring this kind of system and from my preliminary results, the performance improvement is significant.

> Converting the main drawing canvas would presumably involve creating a new offscreen canvas and doing all of the tile drawing on it, then copying the result to the onscreen canvas. I have no idea if that would be faster (because the bulk of the drawing is happening offscreen) or slower (because there's an additional draw call to go from the offscreen canvas to the onscreen one); I guess one would just need to try it and see. It certainly would increase our memory requirement, which wouldn't be welcome in some scenarios.

With offscreen canvas, you don't do any manual copying so I'm not sure what you're exactly referring to. You simply convert a normal canvas element to an OffscreenCanvas and pass that to a WebWorker. The worker does all the rendering on the canvas and the results are immediately available in the browser.

> I'm not sure what we would use WebWorkers for... We want the drawing to happen on the main thread, because our goal is to do it 60 frames a second and shuttling data back and forth between workers will introduce latency. The processing that would be nice to move off the main thread is loading the tiles, but I think the big weight there is decompressing the image files as they load, and I don't think that can be done outside the main thread (unless we want to include our own image decoders, which I'm pretty sure we don't).

The trick is that the network fetching is done on the WebWorker as well. In my test, I didnt use any messaging between the main thread and the worker at all. I only created a SharedArrayBuffer that encoded user movements like zoom, pan etc and set a dirty bit. The main rendering loop in the worker would then check if dirty bit is set and recalculate the viewport based on that. I think in total it was an array of like 6 integers that needed to be shared between the main thread and the worker. Super efficient and no message passing at all.

**Comment by Aiosa** [comment]
Yes, off screen rendering is indeed a good approach I wanted to sooner or later introduce. But the main advantage of that approach is that you use this renderer to all rendering operations which prevents the problem of running out of webgl contexts. There is not much saving on performance, except that you can use all the data once, there is no need to upload textures for navigator and main viewport separately.

Web workers are Indeed a performance improvement, but using a shared array to pass around few numbers is not worth the effort tbh. Moreover, shared array has strict usage policies which are often not met - it's enough to just use CDN to violate them. Shared array would be a good thing for the render data transfer, but... The policies. They hurt.

Note that you don't need to check for any 'dirty' flag moments, osd is doing all the work for you and draw or fetch methods are called when they are necessary.

**Comment by nmzali** [comment]
I understand the strict usage policies around SharedArrayBuffers so for the few things like telling the worker the current zoom, pan, etc this can be done via message passing.

As for the dirty flag thing, my personal project doesn't use OSD. I was building from scratch and thats what I decided on for my implementation. So yeah it wasn't interfacing with OSD at all.

**Comment by Aiosa** [comment]
Actually forget about the CDN, just fetching tiles from extrernal source counts, IMHO quite a frequent usecase. Of course you can do _if_ usage, but that will almost never be used :/  [We have also official webgl renderer here](https://github.com/openseadragon/flex-render) that builds on webgl2 and attempts to do more than what basic renderers do, it's also a place where it's easier to test new stuff rather than in the main library, so if you want to try to implement something similar here or on the master OSD repo, feel free to open a PR. Just note that it will be a lengthy process...

**Comment by nmzali** [comment]
I'll look into it when I have some more free time. I need to get a bit familiar with the OSD codebase and see how I'd fit this in. I'll let you know once I take a look.

**Comment by iangilman** [comment]
We're always interested in performance improvements! But we also want to keep the complexity down as much as possible. Anyway, I'm all for exploring the possibilities 😊

**Comment by Aiosa** [comment]
Yes, that's also why advanced drawers can be added externally. But, with webGL, you probably need to use the offscreen renderer simply because putting multiple viewers on a single page can crash the page - not a good experience.
